src/Decompiler/Scanning/VarargsFormatScanner.cs

```diff
@@ -0,0 +1,189 @@
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Reko.Core;
+using Reko.Core.CLanguage;
+using Reko.Core.Code;
+using Reko.Core.Expressions;
+using Reko.Core.Serialization;
+using Reko.Core.Types;
+using Reko.Evaluation;
+
+namespace Reko.Scanning
+{
+    /// <summary>
+    /// Try to read varargs format, then to parse it, get arguments types and
+    /// build application instruction.
+    /// </summary>
+    class VarargsFormatScanner
+    {
+        private Program program;
+        private IProcessorArchitecture arch;
+        private Frame frame;
+        private ExpressionSimplifier eval;
+        private FunctionType sig;
+
+        public VarargsFormatScanner(
+            Program program,
+            Frame frame,
+            EvaluationContext ctx)
+        {
+            this.program = program;
+            this.arch = program.Architecture;
+            this.frame = frame;
+            this.eval = new ExpressionSimplifier(ctx);
+        }
+
+        public Instruction BuildInstruction(Expression callee, CallSite site)
+        {
+            var pc = callee as ProcedureConstant;
+            if (pc != null)
+                pc.Procedure.Signature = this.sig;
+            var ab = CreateApplicationBuilder(callee, this.sig, site);
+            return ab.CreateInstruction();
+        }
+
+        public bool TryScan(FunctionType sig, ProcedureCharacteristics chr)
+        {
+            if (
+                sig == null || !sig.IsVarargs() ||
+                chr == null || !VarargsParserSet(chr)
+            )
+            {
+                this.sig = null;
+                return false;
+            }
+            var format = ReadVarargsFormat(sig);
+            var argTypes = ParseVarargsFormat(chr, format, program.Platform);
+            this.sig = ReplaceVarargs(sig, argTypes);
+            return true;
+        }
+
+        private ApplicationBuilder CreateApplicationBuilder(
+            Expression callee,
+            FunctionType sig,
+            CallSite site)
+        {
+            var ab = new ApplicationBuilder(
+                arch,
+                frame,
+                site,
+                callee,
+                sig,
+                false);
+            return ab;
+        }
+
+        private Expression GetValue(Expression op)
+        {
+            return op.Accept<Expression>(eval);
+        }
+
+        private string ReadVarargsFormat(FunctionType sig)
+        {
+            var formatIndex = sig.Parameters.Length - 2;
+            if (formatIndex < 0)
+                throw new ApplicationException(
+                    string.Format("Varargs: should be at least 2 parameters"));
+            var formatParam = sig.Parameters[formatIndex];
+            var stackStorage = formatParam.Storage as StackStorage;
```

Comment:
What this means is a `//$TODO` and a (disabled) unit test that simulates varargs on a RISC machine, where the first, say, 3 parameters are passed in registers.